packages/astro/src/core/dev/index.ts

```diff
@@ -277,10 +277,22 @@ export class AstroDevServer {
     const reqStart = performance.now();
     let filePath: URL | undefined;
     try {
-      let routePathname = pathname.startsWith(this.pathname) ? pathname.substr(this.pathname.length) || '/' : undefined;
-      if (!routePathname) {
-        next();
-        return;
+      let routePathname: string = pathname;
+      // If using a subpath, ensure that the user has included the pathname
+      // such as /blog in the URL.
+      if(this.devRoot !== '/') {
+        if(pathname.startsWith(this.devRoot)) {
+          // This includes the subpath, so strip off the subpath so that
+          // matchRoute finds this route.
+          routePathname = pathname.substr(this.devRoot.length) || '';
```

Comment:
mentioned in chat: we have utility functions already to add/remove trailing/leading slashes.